src/charm.py

```diff
@@ -3,58 +3,116 @@
+            if self.config["path"]:
+                self._stored.path = self.config["path"]
+
+        if not (self.config["token"] and self._stored.path):
+            self.unit.status = BlockedStatus("Waiting for Token and Path config")
             return
-        self.unit.status = MaintenanceStatus("Registering runner")
-        token = event.params["token"]
-        url = event.params["url"]
-        if not token:
-            event.fail("No token provided")
+
+        self._reconcile_runners()
+        self._add_cron("check-runners", self.config["check-interval"])
+
+    def _on_check_runners(self, event):
+        if not (self.config["token"] and self._stored.path):
+            logger.info("Not checking runners, missing config")
             return
-        if not url:
-            event.fail("No url provided")
+        self._reconcile_runners()
+
+    def _on_check_runners_action(self, event):
+        if not (self.config["token"] and self._stored.path):
+            event.set_results({"status": "Check aborted, config options missing"})
             return
-        try:
-            self._runner.register(url, token)
-        except subprocess.CalledProcessError as e:
-            logger.error("Register: failed")
-            print(e)
-            logger.error(f"Register: return code {e.returncode}")
-            logger.error(f"Register: output {e.output}")
-            raise
-        event.set_results({"status": "runner registered"})
+
+        self._reconcile_runners()
+        event.set_results({"status": "Completed runner check"})
+
+    def _reconcile_runners(self):
+        delta = self.config["quantity"] - self._runner.active_count()
+        logger.info(f"Reconciling {delta} runners")
+        self._runner.remove_runners()
+        while delta > 0:
+            self.unit.status = MaintenanceStatus(f"Installing {delta} runner(s)")
+            try:
+                self._runner.create(image="ubuntu", virt=self.config["virt-type"])
+            except RuntimeError as e:
+                logger.error("Failed to create runner")
+                logger.error(f"Error: {e}")
+            delta = self.config["quantity"] - self._runner.active_count()
         self.unit.status = ActiveStatus("Active and registered")
-        self._stored.registered = True
+
+    def _on_stop(self, event):
+        self._remove_cron("check-runners")
+        self._runner.remove_runners()
+
+    def _add_cron(self, event, interval):
+        """Add a cron job for the provided event to run at the provided interval."""
+        root_cron = self._cron_tab(user="root")
+        comment = f"Charm cron for {event}"
+
+        comments = [c for c in root_cron.find_comment(comment)]
+        if comments:
+            return  # Already exists
+
+        unit = self.unit.name
+        dispatch = str(self.charm_dir / "dispatch")
+        command = f"juju-run {unit} 'JUJU_DISPATCH_PATH={event} {dispatch}'"
+        job = root_cron.new(command=command, comment=comment)
```

Comment:
As written, it won't update the interval if it changes.

```suggestion
        try:
            job = next(root_cron.find_comment(comment))
        except StopIteration:
            unit = self.unit.name
            dispatch = str(self.charm_dir / "dispatch")
            command = f"juju-run {unit} 'JUJU_DISPATCH_PATH={event} {dispatch}'"
            job = root_cron.new(command=command, comment=comment)
```